lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -4932,6 +4930,43 @@ namespace Js
         }
     }
 
+    bool JavascriptOperators::CanShortcutOnUnknownPropertyName(RecyclableObject *instance)
+    {
+        if (!CanShortcutInstanceOnUnknownPropertyName(instance))
+        {
+            return false;
+        }
+        return CanShortcutPrototypeChainOnUnknownPropertyName(instance->GetPrototype());
+    }
+
+    bool JavascriptOperators::CanShortcutInstanceOnUnknownPropertyName(RecyclableObject *instance)
+    {
+        if (JavascriptProxy::Is(instance))
+        {
+            return false;
+        }
+        if (!instance->HasDeferredTypeHandler())
+        {
+            return true;
+        }
+        return !(instance->IsExternal() || 
```

Comment:
External objects with deferred type handlers do exist. They’re the case that’s the source of the bug. The DOM has property names that the JS engine hasn’t seen, and if the type handler is deferred then the external object may have one of those unknown property names.

--Paul

From: Ian Halliday [mailto:notifications@github.com]
Sent: Friday, April 8, 2016 3:28 PM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>
Subject: Re: [Microsoft/ChakraCore] Fix optimized element access that incorrectly reports that the instance has no such property (#723)


In lib/Runtime/Language/JavascriptOperators.cpp<https://github.com/Microsoft/ChakraCore/pull/723#discussion_r59097941>:

> +            return false;

> +        }

> +        return CanShortcutPrototypeChainOnUnknownPropertyName(instance->GetPrototype());

> +    }

> +

> +    bool JavascriptOperators::CanShortcutInstanceOnUnknownPropertyName(RecyclableObject *instance)

> +    {

> +        if (JavascriptProxy::Is(instance))

> +        {

> +            return false;

> +        }

> +        if (!instance->HasDeferredTypeHandler())

> +        {

> +            return true;

> +        }

> +        return !(instance->IsExternal() ||

I still think this is wrong. IsExternal is orthogonal to HasDeferredTypeHandler and I believe external objects will always have non-deferred type handlers, so IsExternal implies !HasDeferredTypeHander which means this will return true above when IsExternal is true. Is that not the case?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/723/files/e214f5f2f6522e53d73613af095ab96ade7c5421#r59097941>